### DISCUSSION TITLE
How can I call client functions from server?

### DISCUSSION THREAD

**Discussion by kamyar1979** [icerpc/icerpc-csharp#3943] [Q&A] [answered]
What I need is bidirectional RPC in which I can call client functions from server as well as calling server from client. Is there a solution other than mixing client and server code?

**Comment by pepone** [comment, marked as answer]
Hi Kamyar,

With IceRPC we prefer the term "server-to-client" or "the other way around" to refer to RPCs initiated by the server, instead of using "bidirectional", with that out of the way let me explain how this works with IceRPC.

With IceRPC you send a request by using an [invoker], and you process a request and return the corresponding response by using a [dispatcher].

On the client side, you typically use the [ClientConnection] or [ConnectionCache] invokers for sending requests to a server, and that is what most examples show:

```
await using var connection = new ClientConnection(new Uri("icerpc://localhost"));
// GreeterProxy is a struct generated by the Slice compiler.
var greeterProxy = new GreeterProxy(connection);
string greeting = await greeterProxy.GreetAsync(Environment.UserName);
```

Here we create a client connection, and we use it as the invoker of the proxy, nothing fancy.

Sending a request the other way around is the same, you just need to use a different invoker. In this case, you would want to use the invoker provided by the [connection context].

This context is made available to the dispatcher through the [IDispatchInformationFeature], the Router's [UseDispatchInformation] extension method adds a middleware to the dispatch pipeline that adds this feature to the requests it dispatches.

With the router correctly configured you can access this invoker using something like:

```
public ValueTask<string> GreetAsync(string name, IFeatureCollection features, CancellationToken cancellationToken)
{
    IDispatchInformationFeature? dispatchInfo = features.Get<IDispatchInformationFeature>();

    // We installed the dispatch middleware in Router to get this dispatch info:
    Debug.Assert(dispatchInfo is not null);
    var greeterProxy = new GreeterProxy(dispatchInfo.ConnectionContext.Invoker);

    string greeting = await greeterProxy.GreetAsync(Environment.UserName);
}
```

The client configured dispatcher will receive the request, you configure the client dispatcher with the [Dispatcher property] of the connection options when you create the client connection.

You might be interested in the new [Thermostat example], which shows with real code how all this works.

[invoker]: https://docs.icerpc.dev/icerpc/invocation/invocation-pipeline#the-invoker-abstraction
[dispatcher]: https://docs.icerpc.dev/icerpc/dispatch/dispatch-pipeline#the-dispatcher-abstraction
[ClientConnection]: https://docs.icerpc.dev/api/csharp/api/IceRpc.ClientConnection.html
[ConnectionCache]:https://docs.icerpc.dev/api/csharp/api/IceRpc.ConnectionCache.html
[connection context]: https://docs.icerpc.dev/api/csharp/api/IceRpc.IConnectionContext.html
[IDispatchInformationFeature]: https://docs.icerpc.dev/api/csharp/api/IceRpc.Features.IDispatchInformationFeature.html
[UseDispatchInformation]: https://docs.icerpc.dev/api/csharp/api/IceRpc.RouterExtensions.html#IceRpc_RouterExtensions_UseDispatchInformation_IceRpc_Router_
[Dispatcher property]: https://docs.icerpc.dev/api/csharp/api/IceRpc.ConnectionOptions.html#IceRpc_ConnectionOptions_Dispatcher
[Thermostat example]: https://github.com/icerpc/icerpc-csharp/tree/main/examples/Thermostat
